src/main/java/duke/util/Program.java

```diff
@@ -0,0 +1,249 @@
+    public static final String INVALID_RANGE_ERROR = "The index you specified is outside the size of the list";
+
+    public Program() throws IOException {
+        this.numItems = 0;
+        loadSavedTasks();
+    }
+
+    StorageDataParser parser = new StorageDataParser();
+    public static int getNumItems() {
+        return numItems;
+    }
+
+    //function takes in an input string from the user, parses it and runs the corresponding function
+    public void parseAndExecuteTask(String string) throws HalException, IOException {
+        if (Objects.equals(string, "list")) {
+            listAllTasks();
+        } else if (Objects.equals(string, "bye")) {
+            this.executeBye();
+        } else if (string.startsWith("done")) {
+            this.executeDoneTask(string);
+        } else if (string.startsWith("deadline")) {
+            addDeadlineTask(string);
+        } else if (string.startsWith("event")) {
+            addEventTask(string);
+        } else if (string.startsWith("todo")){
+            addToDoTask(string);
+        } else if (string.startsWith("delete")) {
+            deleteTask(string);
+        } else {
+            throw new HalException("I'm sorry, but I don't know what that means :((");
+        }
+        UserData.writeToFile(parser.saveListAsString(listTasks));
+    }
+
+    //function to load tasks from memory
+    public static void loadSavedTasks() throws IOException {
+        listTasks = readFromFile();
+        int count = 0;
+        for (Object obj : listTasks) {
+            if ( obj != null ) {
+                count++;
+            }
+        }
+        numItems = count;
+    }
+
+    //function to add a new deadline task
+    public static void addDeadlineTask(String deadlineTask) throws HalException {
+        System.out.println(LINE_BREAK_SINGLE);
+
+        //check if string contains '/by' tag
+        if (!deadlineTask.contains("/by")) {
+            System.out.println(PRINT_ERROR_MESSAGE);
+            throw new HalException("Wrong Deadline task format");
+        }
+
+        String description = deadlineTask.substring(DEADLINE_INDEX, deadlineTask.indexOf('/')).trim();
+        String by = deadlineTask.substring(deadlineTask.indexOf("/by") + TASK_STRING_OFFSET).trim();
+
+        if (description.equals("")) {
+            System.out.println(PRINT_EMPTY_DESCRIPTION_MESSAGE);
+            throw new HalException("Empty description");
+        } else if (by.equals("")) {
+            System.out.println(PRINT_EMPTY_DATE_MESSAGE);
+            throw new HalException("Empty date");
+        }
+
+        Deadline newDeadlineTask = new Deadline(description, by);
+        listTasks.add(newDeadlineTask);
+        numItems++;
+
+        System.out.println(TASK_ADDED_DONE_TEXT);
+        System.out.println(newDeadlineTask.toString());
+        printTotalTasks();
+        System.out.println(LINE_BREAK_SINGLE);
+        System.out.print(ENTER_COMMAND_TEXT);
+    }
```

Comment:
Method seems a bit long and parts of it are similarly used in other methods. Maybe it would be good to make use of SLAP to improve quality